tests/helpers/deviceappproxy/apduabstract.py

```diff
@@ -0,0 +1,129 @@
+from typing import List, Dict, Optional, cast, NewType, Tuple, AnyStr
+from dataclasses import dataclass, field
+
+
+@dataclass
+class CApdu:
+    class Type:
+        IN: int = 0
+        OUT: int = 1
+        INOUT: int = 2
+    data: List[bytes]
+    cla: AnyStr = field(default="00")
+    ins: AnyStr = field(default="00")
+    p1: AnyStr = field(default="00")
+    p2: AnyStr = field(default="00")
+    lc: AnyStr = field(default="00")
+    le: AnyStr = field(default="00")
+    typ: Type = field(default=Type.INOUT)
+
+
+ApduDict = NewType("ApduDict", Dict[str, CApdu])
+
+
+class ApduSet:
+    """Collects a set of CApdu objects and provides their raw version, ready
+    to be sent to a device/an app"
+    """
+    _apdus: ApduDict = {}
+
+    def __init__(self, apdus: Optional[ApduDict] = None, max_lc: int = 255) -> None:
+        self.apdus = apdus
+        self.max_lc = max_lc
+
+    @staticmethod
+    def _bytes(data: AnyStr) -> bytes:
+        if isinstance(data, (bytes, bytearray)):
+            return data
+        if isinstance(data, int):
+            return bytes([cast(int, data)])
+        if isinstance(data, str):
+            return bytes.fromhex(data)
+        raise TypeError(f"{data} cannot be converted to bytes")
+
+    @property
+    def apdus(self) -> Optional[ApduDict]:
+        return ApduSet._apdus if len(ApduSet._apdus.keys()) > 0 else None
+
+    @apdus.setter
+    def apdus(self, new_apdus: ApduDict, overwrite: bool = False) -> None:
+        """Sets a new CApsu internal dictionary if it wasn't set at instanciation time,
+        unless overwrite is True."""
+        if not self.apdus or overwrite is True:
+            if not isinstance(new_apdus, dict):
+                raise ValueError("Attribute newApdus must be a dictionary containing CApdu "
+                                 "instances as values")
+            ApduSet._apdus = new_apdus
+
+    def apdu(self, name: str,
+             p1: Optional[AnyStr] = None,
+             p2: Optional[AnyStr] = None,
+             data: Optional[List[AnyStr]] = None,
+             le: Optional[AnyStr] = None) -> Tuple[bytes, List[Optional[bytes]]]:
+        """Returns the raw bytes for the C-APDU header requested by name.
```

Comment:
It seems to return more than just the raw bytes, right? what's that list that it returns as well?

Reply:
and this method actually changes the adpu on the fly as well? This seems wrong, those are two different actions that should be split

Reply:
It used to return the raw bytes as a single bytes buffer. But I've changed the implementation to follow the changes in the `DeviceAppProxy.send_apdu()` method (which actually communicates with the Nano app). Now this function returns  a tuple of of elements ready to be unpacked and passed as parameters to the `DeviceAppProxy.send_apdu()` method. 
==> I've updated the docstring to reflect that behaviour.

The function changes only the dynamic (i.e. that change each time the APDU is sent to the secure app, according to the app's spec) parts of the APDU stored in the `CApdu` object: p1, p2, payload data attributes to the values of the APDU being sent and computes automatically the correct values for lc and le. I could have made a more distinct separation between the static parts (cla, ins) and the dynamic parts of the APDU in the definition of CApdu but I don't think it's worth it at the moment.

Reply:
👍